benchmark/single-source/PrimsNonStrongRef.swift

```diff
@@ -0,0 +1,1109 @@
+  (33, 79, 1871),
+  (76, 73, 514),
+  (98, 37, 1102),
+  (45, 44, 364),
+  (16, 69, 1183),
+  (49, 41, 1629),
+  (19, 80, 1838),
+  (71, 57, 500),
+  (6, 4, 634),
+  (64, 27, 384),
+  (84, 86, 1431),
+  (5, 17, 633),
+  (96, 88, 334),
+  (87, 5, 190),
+  (70, 21, 1619),
+  (55, 33, 1391),
+  (10, 63, 1681),
+  (11, 62, 1339),
+  (33, 13, 826),
+  (64, 70, 878),
+  (65, 72, 1330),
+  (70, 55, 1264),
+  (64, 55, 1996),
+  (50, 41, 1559),
+  (46, 99, 1576),
+  (88, 96, 334),
+  (51, 20, 868),
+  (73, 7, 754),
+  (80, 70, 995),
+  (44, 84, 952),
+  (29, 19, 883),
+  (59, 69, 903),
+  (57, 53, 1575),
+  (90, 13, 1030),
+  (28, 38, 414),
+  (12, 60, 1240),
+  (85, 58, 573),
+  (90, 55, 1808),
+  (4, 10, 1808),
+  (68, 44, 1229),
+  (92, 33, 946),
+  (90, 81, 1024),
+  (53, 75, 545),
+  (45, 30, 272),
+  (41, 77, 138),
+  (21, 70, 1619),
+  (45, 73, 1876),
+  (35, 68, 1252),
+  (13, 96, 461),
+  (53, 57, 1575),
+  (82, 89, 409),
+  (28, 61, 449),
+  (58, 61, 78),
+  (27, 80, 1010),
+  (61, 58, 78),
+  (38, 36, 529),
+  (80, 30, 397),
+  (18, 59, 1352),
+  (62, 11, 1339),
+  (95, 59, 285),
+  (51, 98, 1160),
+  (6, 18, 696),
+  (30, 80, 397),
+  (69, 94, 1397),
+  (58, 85, 573),
+  (48, 99, 410),
+  (51, 46, 213),
+  (57, 71, 500),
+  (91, 30, 104),
+  (65, 7, 349),
+  (79, 51, 175),
+  (47, 26, 921),
+  (4, 61, 543),
+  (98, 73, 1749),
+  (74, 77, 1277),
+  (61, 28, 449),
+  (58, 8, 1131),
+  (61, 45, 1608),
+  (74, 87, 934),
+  (71, 29, 1788),
+  (30, 91, 104),
+  (13, 1, 1187),
+  (0, 26, 1993),
+  (82, 49, 1292),
+  (43, 87, 579),
+  (24, 45, 1720),
+  (20, 51, 868),
+  (77, 62, 907),
+  (82, 75, 306),
+]
+
+struct PrimsState<Node: GraphNode, Box> where Node.BoxType == Box, Box.ValueType == Node {
+  var graph: Array<Node>
+  var edgeToCostMap: Dictionary<Edge<Node>, Double>
+
+  init(_ _: Node.Type) {
+    // Prepare graph and edge->cost map
+    graph = Array<Node>()
+    for nodeId in nodes {
+      graph.append(Node(id: nodeId))
+    }
+
+    edgeToCostMap = Dictionary<Edge<Node>, Double>()
+    for tup in edges {
+      edgeToCostMap[Edge(start: Box(graph[tup.0]), end: Box(graph[tup.1]))] = tup.2
+      graph[tup.0].adjList.append(Box(graph[tup.1]))
+    }
+  }
+}
+
+let weakPrimsState = PrimsState(WeakGraphNode.self)
+let unownedSafePrimsState = PrimsState(UnownedSafeGraphNode.self)
+let unownedUnsafePrimsState = PrimsState(UnownedUnsafeGraphNode.self)
+let unmanagedPrimsState = PrimsState(UnmanagedGraphNode.self)
+
+//===----------------------------------------------------------------------===//
+//                                 Protocols
+//===----------------------------------------------------------------------===//
+
+protocol ValueBox : Hashable {
+  associatedtype ValueType : AnyObject
+
+  init(_ inputValue: ValueType)
+  var value: ValueType { get }
+
+  func free()
+}
+
+extension ValueBox {
+  func free() {}
+}
+
+protocol GraphNode : AnyObject {
+  associatedtype BoxType : ValueBox where BoxType.ValueType : GraphNode
```

Comment:
Would it make sense for the where clause to be `where BoxType.ValueType == Self`?

Reply:
I can fix that. Thanks!

Reply:
Done. Allowed me to simplify a bunch of other constraints actually.